File: src/rematch/library_interface/match.cpp
#include "match.hpp"

#include "evaluation/document.hpp"

using namespace rematch;

namespace REMatch {
inline namespace library_interface {

Match::Match(
    rematch::mediator::Mapping mapping,
    std::shared_ptr<rematch::parsing::VariableCatalog> variable_catalog,
    std::shared_ptr<Document> document)
    : mapping_(mapping),
      variable_catalog_(variable_catalog),
      document_(document) {}

int Match::start(std::string variable_name) {
  auto const span = this->span(variable_name);
  return span.first;
}

int Match::start(int variable_id) {
  return start(variable_catalog_->get_var(variable_id));
}

int Match::end(std::string variable_name) {
  auto const span = this->span(variable_name);
  return span.second;
}

int Match::end(int variable_id) {
  return end(variable_catalog_->get_var(variable_id));
}

Span Match::span(std::string variable_name) {
  return mapping_.get_span_of_variable(variable_name);
}

Span Match::span(int variable_id) {
  return span(variable_catalog_->get_var(variable_id));
}

std::string Match::group(std::string variable_name) {
  const auto span = this->span(variable_name);
  return document_->substr(span.first, span.second - span.first);
}

std::string Match::group(int variable_id) {
  return group(variable_catalog_->get_var(variable_id));
}

std::map<std::string, Span> Match::groupdict() {
  return mapping_.get_spans_map();
}

std::vector<std::string> Match::variables() {
  if (empty())
    return {};

  return variable_catalog_->variables();
}

bool Match::empty() {
  return mapping_.get_spans_map().size() == 0;
}

std::ostream& operator<<(std::ostream& os, Match& match) {
  int amount_of_variables = match.variable_catalog_->size();

  for (unsigned int i = 0; i < match.variable_catalog_->size() - 1; i++) {
    std::string variable_name = match.variable_catalog_->get_var(i);
    Span span = match.span(variable_name);
    os << variable_name << " = |" << span.first << "," << span.second << ">\t";
  }

  std::string variable_name = match.variable_catalog_->get_var(amount_of_variables - 1);
  Span span = match.span(variable_name);
  os << variable_name << " = |" << span.first << "," << span.second << ">";

  return os;
}

}  // namespace library_interface
}  // namespace REMatch

File: src/targets/java/rematch/Match.java
package rematch;

import java.util.HashMap;
import java.util.Map;

public class Match {
    private final javacpp.Match cppMatch;

    public Match(javacpp.Match cppMatch) {
        this.cppMatch = cppMatch;
    }

    public int start(String variableName) {
        return (int) this.cppMatch.start(variableName);
    }

    public int start(int variableId) {
        return (int) this.cppMatch.start(variableId);
    }

    public int end(String variableName) {
        return (int) this.cppMatch.end(variableName);
    }

    public int end(int variableId) {
        return (int) this.cppMatch.end(variableId);
    }

    public String group(String variableName) {
        return this.cppMatch.group(variableName);
    }

    public String group(int variableId) {
        return this.cppMatch.group(variableId);
    }

    public javacpp.Span span(int variableId) {
        return this.cppMatch.span(variableId);
    }

    public javacpp.Span span(String variableName) {
        return this.cppMatch.span(variableName);
    }

   
    public String[] variables() {
        javacpp.StringVector cppVariables = this.cppMatch.variables();
        String[] variables = new String[(int) cppVariables.size()];

        for (int i = 0; i < cppVariables.size(); i++) {
            variables[i] = cppVariables.at(i);
        }

        return variables;
    }

    public Map<String, javacpp.Span> groupdict() {
        javacpp.StringSpanMap cppGroupDict = this.cppMatch.groupdict();
        javacpp.StringVector vars = this.cppMatch.variables();
        Map<String, javacpp.Span> groupDict = new HashMap<>();
        
        for (long i = 0; i < vars.size(); i++) {
            String key = vars.at(i).toString();
            javacpp.Span span = cppGroupDict.at(key);
            groupDict.put(key, span);
        }
        //  result.groupdict().at("x").first() + ", " + result.groupdict().at("x").second());

        return groupDict;
    }




    public boolean empty() {
        return this.cppMatch.empty();
    }

    public String toString() {
        return this.cppMatch.to_string();
    }

    
}
